cirq/sim/state.py

```diff
@@ -121,3 +123,157 @@ def validate_normalized_state(state: np.ndarray, num_qubits: int,
+            a bitstring, has the largest values that the 0th index.
+        repetitions: The number of times to sample the state.
+
+    Returns:
+        Measurement results with True corresponding to the |1> state.
+        The outer list is for repetitions, and the inner corresponds to
+        measurements ordered by the input indices.
+
+    Raises:
+        ValueError if repetitions is less than one or size of state is not a
+            power of 2.
+        IndexError if the indices are out of range for the number of qubits
+            corresponding to the state.
+    """
+    if repetitions < 1:
+        raise ValueError(
+            'Number of repetitions cannot be negative. Was {}'.format(
+                repetitions))
+    num_qubits = _validate_num_qubits(state)
+    _validate_indices(num_qubits, indices)
+
+    if len(indices) == 0:
+        return [[]]
+
+    # Calculate the measurement probabilities.
+    probs = _probs(state, indices, num_qubits)
+
+    # We now have the probability vector, correctly ordered, so sample over
+    # it. Note that we us ints here, since numpy's choice does not allow for
+    # choosing from a list of tuples or list of lists.
+    result = np.random.choice(len(probs), size=repetitions, p=probs)
+    # Convert to bools and rearrange to match repetition being the outer list.
+    return np.transpose([(1 & (result >> i)).astype(np.bool) for i in
+                         range(len(indices))]).tolist()
+
+
+def measure_state_vector(
+        state: np.ndarray,
+        indices: List[int],
+        out: np.ndarray = None) -> Tuple[List[bool], np.ndarray]:
+    """Performs a measurement of the state in the computational basis.
+
+    This does not modify `state` unless the optional `out` is `state`.
+
+    Args:
+        state: The state to be measured. This state is assumed to be normalized.
+            The state must be of size 2 ** integer.  The state can be of shape
+            (2 ** integer) or (2, 2, ..., 2).
+        indices: Which qubits are measured. The state is assumed to be supplied
+            in big endian order. That is the xth index of v, when expressed as
+            a bitstring, has the largest values in the 0th index.
+        out: An optional place to store the result. If out is the same as
+            the `state` parameter, then state will be modified inline. If out
+            is not None, then the result is put into out.  If out is None
+            a new value will be allocated. In all of these case out will be the
+            same as the returned ndarray of the method. The shape and dtype of
+            out will always match that of state.
+
+    Returns:
+        A tuple of a list and an numpy array. The list is an array of booleans
+        corresponding to the measurement values (ordered by the indices). The
+        numpy array is the post measurement state. This state has the same
+        shape and dtype as the input state.
```

Comment:
Actually, it now matches the shape and dtype of `state if out is None else out`.

Reply:
done